### PR TITLE
send Invalidate cache of subscription/s through websocket

### PR DESCRIPTION
**Pull request by tjeerddie**
- invalidate subscription/s cache in cache domain model step.
- invalidate subscription/s cache in 'set in sync' endpoint.
- invalidate subscription/s cache in create and update subscription description.
- Change invalidation to return a dict instead of a list to have it in line with frontend cache invalidation, [check the discussion](https://github.com/workfloworchestrator/orchestrator-core/pull/671#discussion_r1628930882)

Closes: https://github.com/workfloworchestrator/orchestrator-core/issues/574

### CODE REVIEW

**Review by Josephine-Rutten — APPROVED** [review]

File: orchestrator/workflows/steps.py [hunk 1]
```diff
@@ -206,10 +207,12 @@ def _cache_other_subscriptions(product_block: ProductBlockModel) -> None:
     for subscription_id in cached_subscription_ids:
         subscription_model = SubscriptionModel.from_subscription(subscription_id)
         to_redis(build_extended_domain_model(subscription_model))
+        sync_broadcast_invalidate_cache(str(subscription.subscription_id))
```
Comment (dmarjenburgh) [inline orchestrator/workflows/steps.py:210]: So, are we only sending a UUID to the frontend? How does the frontend know it's about a subscription?

Comment (tjeerddie) [reply orchestrator/workflows/steps.py]: yea, good spot...

changed to `subscription:uuid` and also changed process to `process:uuid`

File: orchestrator/websocket/__init__.py [hunk 2]
```diff
@@ -118,7 +130,7 @@ def send_process_data_to_websocket(
         logger.debug("Broadcast process data directly to websocket_manager", process_id=str(process_id))
 
         anyio.run(websocket_manager.broadcast_data, [WS_CHANNELS.ALL_PROCESSES], data)
-        sync_broadcast_invalidate_cache("processes", str(process_id))
+        sync_broadcast_invalidate_cache("processes", f"process:{process_id}")
```
Comment (Mark90) [inline orchestrator/websocket/__init__.py:133]: As far as I remember we were using arrays as cache keys, so something like

```
["processes", <uuid>]
["subscriptions", <uuid>]
```

Did that format change to colon-separated strings now? 

@DutchBen to verify from frontend perspective

Comment (dmarjenburgh) [reply orchestrator/websocket/__init__.py:133]: An array as cache key sounds like a bad design choice. Are you sure it's not for sending multiple cache keys in a single send?

Comment (Mark90) [reply orchestrator/websocket/__init__.py:133]: It's not meant for sending multiple keys, just an alternative to a delimited string. What's particularly bad about it?

I don't have a preference for one or the other, I'm fine withing a string delimited by `:` or `.` or something else. We'd just have to do this in accordance with the UI.

Comment (DutchBen) [reply orchestrator/websocket/__init__.py:133]: I was interpreting the message ["processes", <uuid>] as one message and the uuid as a processes uuid. I understand from @tjeerddie that it is in fact 2 messages 'processes' and a uuid that is not necessarily a process UUID. I can go either way with a slight preference for one message per message.

Comment (dmarjenburgh) [reply orchestrator/websocket/__init__.py:133]: Cache keys need to be checked for equality and checking arrays for equality in JS can easily lead to ambiguities, since they are not values like strings are.

Or, if you stringify them to JSON, is `["process","123"] == ["process", "123"]`?

Comment (tjeerddie) [reply orchestrator/websocket/__init__.py:133]: @DutchBen showed me how we use the caching on the frontend.
which is a object `{ type: 'processes' }` for list and `{ type: 'processes', id: "uuid" }` for a process detail.
I suggest we send it exactly like that, so its consistent on both frontend and backend, removing confusion on how one would add their own cache invalidation with both projects.

and do we keep the list or send one object per websocket message?

Comment (Mark90) [reply orchestrator/websocket/__init__.py:133]: So just to make sure we're on the same page, the messages as a whole currently look like this;

```json
{"name": "invalidateCache", "value": ["processes", "(the id here)"]}

{"name": "invalidateCache", "value": ["engineStatus"]}
```

And we'd want to change them to look like below?

```json
{"name": "invalidateCache", "value": {"type": "processes", "id": "(the id here)"}}

{"name": "invalidateCache", "value": {"type": "engineStatus"}}
```

If so: it's more explicit, I think it's great. 

> and do we keep the list or send one object per websocket message?

I think one object is fine.

Comment (Mark90) [reply orchestrator/websocket/__init__.py:133]: > Cache keys need to be checked for equality and checking arrays for equality in JS can easily lead to ambiguities, since they are not values like strings are.

Fair point, I didn't know this was ambiguous in JS. I agree to change to an object as above.

File: orchestrator/utils/redis.py [hunk 3]
```diff
@@ -86,8 +86,8 @@ def delete_subscription_from_redis(
 ) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
     def _delete_subscription_from_redis(func: Callable[..., Any]) -> Callable[..., Any]:
         @functools.wraps(func)
-        def wrapper(*args: tuple, **kwargs: dict[str, Any]) -> Any:
-            data = func(*args, **kwargs)
+        async def wrapper(*args: tuple, **kwargs: dict[str, Any]) -> Any:
+            data = await func(*args, **kwargs)
```
Comment (Mark90) [inline orchestrator/utils/redis.py:90]: Why is this changed to async?  In surf orchestrator this function is used as a decorator on sync functions. And even if we change those to async, then this will break it for other users
